embedded-graphics/src/primitives/circle.rs

```diff
@@ -206,37 +203,105 @@ where
             -(self.primitive.radius as i32),
             -(self.primitive.radius as i32),
         );
+
+        // A stroke width of zero renders a 1px stroke, so add 1 to inner radius to compensate
+        let inner_radius = self.primitive.radius as i32 - self.style.stroke_width_i32() + 1;
+        let outer_radius = self.primitive.radius as i32;
+
+        let inner_threshold = inner_radius.pow(2) - inner_radius;
+        let mut outer_threshold = outer_radius.pow(2) + outer_radius;
+
+        // Special case for small circles. This kludge removes the top-left pixel and leaves the
+        // circle as a `+` shape.
+        if self.primitive.radius == 1 {
+            outer_threshold -= 1;
+        }
+
         StyledCircleIterator {
             center: self.primitive.center,
             radius: self.primitive.radius,
             style: self.style,
             p: top_left,
+            outer_threshold,
+            inner_threshold,
         }
     }
 }
 
 #[cfg(test)]
 mod tests {
     use super::*;
+    use crate::mock_display::MockDisplay;
     use crate::pixelcolor::BinaryColor;
 
+    // Check that tiny circles render as a "+" shape with a hole in the center
+    #[test]
+    fn tiny_circle() {
+        let mut display = MockDisplay::new();
+
+        Circle::new(Point::new(1, 1), 1)
+            .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+            .draw(&mut display);
+
+        #[rustfmt::skip]
+        assert_eq!(
+            display,
+            MockDisplay::from_pattern(&[
+                " # ",
+                "# #",
+                " # "
+            ])
+        );
+    }
+
+    // Check that tiny filled circle render as a "+" shape with NO hole in the center
+    #[test]
+    fn tiny_circle_filled() {
+        let mut display = MockDisplay::new();
+
+        Circle::new(Point::new(1, 1), 1)
+            .into_styled(PrimitiveStyle::with_fill(BinaryColor::On))
+            .draw(&mut display);
+
+        #[rustfmt::skip]
```

Comment:
Added so `from_pattern(...)` stays on multiple lines. I would've liked to have put this on the `MockDisplay::from_pattern` line but this is the closest I could get without throwing warnings.